plugin/kubernetes/pod.go

```diff
@@ -19,56 +19,98 @@ import (
 
 type pod struct {
 	plugin.EntryBase
-	client    *k8s.Clientset
-	config    *rest.Config
-	ns        string
-	startTime time.Time
+	client *k8s.Clientset
+	config *rest.Config
+	ns     string
 }
 
-func newPod(client *k8s.Clientset, config *rest.Config, ns string, p *corev1.Pod) *pod {
+func newPod(ctx context.Context, client *k8s.Clientset, config *rest.Config, ns string, p *corev1.Pod) (*pod, error) {
 	pd := &pod{
 		EntryBase: plugin.NewEntry(p.Name),
 		client:    client,
 		config:    config,
 		ns:        ns,
-		startTime: p.CreationTimestamp.Time,
 	}
+	pd.DisableDefaultCaching()
 
-	return pd
-}
-
-func (p *pod) Metadata(ctx context.Context) (plugin.MetadataMap, error) {
-	pd, err := p.client.CoreV1().Pods(p.ns).Get(p.Name(), metav1.GetOptions{})
+	pdInfo, err := pd.cachedPodInfo(ctx, false)
 	if err != nil {
 		return nil, err
 	}
+	pdInfo.pd = p
+
+	meta := pdInfo.toMeta()
+	attr := plugin.EntryAttributes{}
+	attr.SetCtime(p.CreationTimestamp.Time)
+	attr.SetAtime(attr.Ctime())
+	attr.SetMeta(meta)
+	pd.SetInitialAttributes(attr)
+	pd.Sync(plugin.SizeAttr(), "LogSize")
 
-	journal.Record(ctx, "Metadata for pod %v: %+v", p.Name(), pd)
-	return plugin.ToMetadata(pd), nil
+	return pd, nil
 }
 
-func (p *pod) Attr(ctx context.Context) (plugin.Attributes, error) {
-	return plugin.Attributes{
-		Ctime: p.startTime,
-		Mtime: time.Now(),
-		Atime: p.startTime,
-		Size:  plugin.SizeUnknown,
-	}, nil
+type podInfoResult struct {
+	pd         *corev1.Pod
+	logContent []byte
 }
 
-func (p *pod) Open(ctx context.Context) (plugin.SizedReader, error) {
-	req := p.client.CoreV1().Pods(p.ns).GetLogs(p.Name(), &corev1.PodLogOptions{})
-	rdr, err := req.Stream()
+func (pdInfo podInfoResult) toMeta() plugin.EntryMetadata {
+	meta := plugin.ToMeta(pdInfo.pd)
+	meta["LogSize"] = len(pdInfo.logContent)
+	return meta
+}
+
+func (p *pod) cachedPodInfo(ctx context.Context, queryPodMeta bool) (podInfoResult, error) {
+	cachedPdInfo, err := plugin.CachedOp("PodInfo", p, 15*time.Second, func() (interface{}, error) {
+		result := podInfoResult{}
+
+		if queryPodMeta {
```

Comment:
This is here to avoid an unnecessary query to `Pods` when the current pod's being created. Note that the result here won't get cached b/c the pod has no id, so this code's safe.

May be worth rethinking this at some point.

Reply:
It's kind of weird code. Is there another way we could refactor it?